File: src/token_storage/src/services/token_price/icexplorer.rs
use std::u128;

use candid::{Nat, Principal};
use ic_cdk::api::management_canister::http_request::{http_request, CanisterHttpRequestArgument, HttpHeader, HttpMethod};
use log::info;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::services::token_price::TokenProvider;

const IC_EXPLORER_BASE_URL:&str = "https://api.icexplorer.io/api";

#[derive(Debug, Serialize, Deserialize)]
struct TokenListRequest {
  pub page: u32,
  pub size: u32
}

// #[derive(Debug, Serialize, Deserialize)]
// struct TokenListRequest {
//   pub page: u32,
//   pub size: u32
// }

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct IcExplorerTokenDetails {
  pub name: String,
  pub symbol: String,
  pub price: Nat,
  pub fee: String,
  pub ledger_id: Principal,
  pub token_decimal: u8,
}

// IcExplorerTokenDetail {
//   controllerArray: string[];
//   cycleBalance: string;
//   fee: string;
//   fullyDilutedMarketCap: string;
//   holderAmount: number;
//   ledgerId: string;
//   marketCap: string;
//   memorySize: string;
//   mintingAccount: string;
//   moduleHash: string;
//   name: string;
//   price: string;
//   priceChange24: string;
//   priceICP: string;
//   source: string;
//   standardArray: string[];
//   supplyCap: string;
//   symbol: string;
//   tokenDecimal: number;
//   totalSupply: string;
//   transactionAmount: number;
//   tvl: string;
//   txVolume24: string;
// }

/// A TokenProvider that fetches token prices from IcExplorer
pub struct IcExplorerTokenProvider {
    pub url: String,
}

impl IcExplorerTokenProvider {
    pub fn new(url: String) -> Self {
        Self { url }
    }
}

impl TokenProvider for IcExplorerTokenProvider {

    fn list_tokens(&self) -> std::pin::Pin<Box<dyn Future<Output = Result<Vec<super::TokenData>, ()>>>> {
Box::pin(async move {

    
    // 2.1 Setup the URL
    let host = "putsreq.com";
    let url = format!("{IC_EXPLORER_BASE_URL}/token/list");
    
    // 2.2 prepare headers for the system http_request call
    //Note that `HttpHeader` is declared in line 4
    let request_headers = vec![
        HttpHeader {
            name: "Content-Type".to_string(),
            value: "application/json".to_string(),
        },
        ];
        
        let json_string = serde_json::to_string(&TokenListRequest {
            page: 1,
            size: 100
        }).unwrap();
        
        let json_utf8: Vec<u8> = json_string.into_bytes();
        let request_body: Option<Vec<u8>> = Some(json_utf8);
        
        let request = CanisterHttpRequestArgument {
            url: url.to_string(),
            max_response_bytes: None, //optional for request
            method: HttpMethod::POST,
            headers: request_headers,
            body: request_body,
            transform: None,
            // transform: None, //optional for request
        };
        
        //3. MAKE HTTPS REQUEST AND WAIT FOR RESPONSE
        
        //Note: in Rust, `http_request()` already sends the cycles needed
        //so no need for explicit Cycles.add() as in Motoko
        match http_request(request, 21_000_000_000).await {
            //4. DECODE AND RETURN THE RESPONSE
            
            //See:https://docs.rs/ic-cdk/latest/ic_cdk/api/management_canister/http_request/struct.HttpResponse.html
            Ok((response,)) => {

                let value: Value = serde_json::from_slice(response.body.as_slice()).unwrap();

                info!("[IcExplorerTokenProvider] values: {:?}", value);

                let list= value.get("data").unwrap().get("list").unwrap();

                let list: Vec<IcExplorerTokenDetails> = serde_json::from_value(list.clone()).unwrap();
                info!("[IcExplorerTokenProvider] list: {:?}", list);
                
                Ok(vec![])
            }
            Err((r, m)) => {
                info!("The http_request resulted into error. RejectionCode: {r:?}, Error: {m}");
                Err(())
            }
        }
    })
    }
}

File: src/cashier_frontend_new/src/modules/token/services/token_price/icExplorer.ts
// Copyright (c) 2025 Cashier Protocol Labs
// Licensed under the MIT License (see LICENSE file in the project root)

import { IC_EXPLORER_BASE_URL } from "$modules/token/constants";
import axios from "axios";
import queryString from "query-string";
import { Err, Ok, type Result } from "ts-results-es";
import { type TokenPriceService } from ".";

/**
 * Simple service for fetching token prices from IC Explorer
 */
class IcExplorerTokenPriceService implements TokenPriceService {
  /**
   * Fetch the list of tokens from IC Explorer
   * @returns Promise resolving to an array of IcExplorerTokenDetail objects
   */
  public async getListToken(): Promise<IcExplorerTokenDetail[]> {
    const url = "token/list";

    const icExplorerAxiosClient = axios.create({
      baseURL: IC_EXPLORER_BASE_URL,
      headers: {
        accept: "application/json",
        "cache-control": "no-cache",
        "content-type": "application/json",
        pragma: "no-cache",
        priority: "u=1, i",
        "Referrer-Policy": "strict-origin-when-cross-origin",
      },
      paramsSerializer: {
        serialize: (params) => queryString.stringify(params),
      },
    });

    const response = await icExplorerAxiosClient.post(url, {
      page: 1,
      size: 300,
    });

    console.log("getListToken response", response.data.data.list);

    return response.data.data.list;
  }

  /**
   * Fetch all token prices from IC Explorer
   * @returns Object mapping token IDs to their USD prices
   */
  public async getTokenPrices(): Promise<
    Result<Record<string, number>, Error>
  > {
    try {
      const client = axios.create({
        baseURL: IC_EXPLORER_BASE_URL,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
        },
        timeout: 10000, // 10 seconds timeout
      });

      const response = await client.post("/token/list", {
        page: 1,
        size: 300, // Fetch up to 300 tokens
      });

      if (response.data.statusCode !== 600) {
        return Err(new Error(`API error: ${response.data.statusCode}`));
      }

      // Map to a simple object of tokenId -> price
      const priceMap: Record<string, number> = {};
      for (const token of response.data.data.list) {
        try {
          priceMap[token.ledgerId] = Number(parseFloat(token.price).toFixed(7));
        } catch (e) {
          console.warn(
            `Failed to parse price for ${token.symbol} (${token.ledgerId})`,
            e,
          );
        }
      }

      return Ok(priceMap);
    } catch (error) {
      return Err(
        new Error(`Failed to fetch prices from IC Explorer: ${error}`),
      );
    }
  }
}

/**
 * Detailed token information from IC Explorer
 */
export interface IcExplorerTokenDetail {
  controllerArray: string[];
  cycleBalance: string;
  fee: string;
  fullyDilutedMarketCap: string;
  holderAmount: number;
  ledgerId: string;
  marketCap: string;
  memorySize: string;
  mintingAccount: string;
  moduleHash: string;
  name: string;
  price: string;
  priceChange24: string;
  priceICP: string;
  source: string;
  standardArray: string[];
  supplyCap: string;
  symbol: string;
  tokenDecimal: number;
  totalSupply: string;
  transactionAmount: number;
  tvl: string;
  txVolume24: string;
}

export const icExplorerTokenPriceService = new IcExplorerTokenPriceService();
